### PR TITLE
Fix for DDP DataLoader Stale Reference Issue in PyTorch Lightning

### PR DESCRIPTION
**Pull request by azadeh-gh**
### Problem Statement
Resolves critical Distributed Data Parallel (DDP) DataLoader stale reference issues that caused training instability, gradient synchronization failures, and memory corruption when using dynamic data windows across multiple GPU ranks.

**Issues Fixed:**

- **Stale Object References**: DataLoaders holding references to outdated data summary dictionaries
- **Rank Desynchronization**: Different DDP ranks accessing inconsistent data windows
- **Memory Corruption**: Reused memory addresses causing object confusion
- **Worker Process Failures**: DataLoader workers accessing outdated object references
- **Training Instability**: Inconsistent data across ranks leading to gradient sync failures
### Solution Overview:
Implements a comprehensive "**Fresh Object Guarantee**" system that ensures all DataLoader objects are completely refreshed on every epoch with new memory addresses, preventing any possibility of stale references in distributed training.

### Key Changes:
**1. Enhanced Callback System** (`callbacks.py`)

- **Before**: Simple callbacks with no DDP awareness or timing safety
- **After**: Production-ready callbacks with comprehensive DDP synchronization
 **New Features:**
✅ DDP-Safe Timing: Updates at epoch boundaries, not during training
✅ Rank Awareness: Full distributed training synchronization
✅ Configurable Validation: Optional validation resampling (default: fixed)
✅ Comprehensive Parameter System: Flexible window management
✅ Error Handling: Robust validation and error recovery

``` python
# Before (Unsafe)
def on_train_epoch_start(self, trainer, pl_module):
    datamodule.setup("fit")  # Dangerous mid-training setup!

# After (DDP-Safe)  
def on_train_epoch_end(self, trainer, pl_module):
    rank = int(os.environ.get("RANK", "0"))
    print(f"[Rank {rank}] [Sampler - TRAIN] Next -> {start.date()} .. {end.date()}")
    self._update_datamodule(trainer, start, end, is_train=True)
```
**2. Redesigned DataModule** (``gnn_datamodule.py``)

- **Before:** Direct object assignment causing stale references
- **After**: Version-managed object lifecycle with fresh creation guarantee
  **New Architecture:**
  ✅ Version Management: Track object freshness with version counters
  ✅ Object Lifecycle Control: Dedicated rebuild methods for fresh objects
  ✅ DDP Worker Management: Rank-aware worker initialization
  ✅ Memory Safety: Comprehensive object ID tracking for debugging
  ✅ Separation of Concerns: Clean train/validation data management
```python
# Before (Stale Reference Risk)
def set_train_data(self, start_date, end_date):
    self.train_data_summary = organize_bins_times(...)  # Same object reference

# After (Fresh Object Guarantee)
def set_train_window(self, start_dt, end_dt):
    self._train_version += 1  # Track updates
    self._rebuild_train_summary()  # Force NEW objects
    print(f"sum_id={id(self.train_data_summary)}")  # Debug tracking
```
**3. Training Configuration Updates** (`train_gnn.py`, `run_gnn.sh`)
  ✅ Enabled reload_dataloaders_every_n_epochs=1 - Essential for dynamic data
  ✅ Enhanced logging for DDP debugging and monitoring
  ✅ Worker safety settings (`persistent_workers=False` during development)

**4. Model & Evaluation Enhancements**
  ✅ First-batch validation logging for debugging data consistency
  ✅ Enhanced error handling for distributed training edge cases
  ✅ Parameter compatibility fixes for seamless integration
### Technical Implementation
```python
def _rebuild_train_summary(self):
    # Always creates NEW objects with NEW memory addresses
    self.train_data_summary = organize_bins_times(...)
    self.train_bin_names = sorted(...)
    print(f"[Rank {rank}] sum_id={id(self.train_data_summary)}")  # Track object IDs

def train_dataloader(self):
    # Every reload gets completely fresh objects
    ds = BinDataset(self.train_bin_names, self.train_data_summary, ...)
    loader = PyGDataLoader(ds, worker_init_fn=self._worker_init, ...)
    print(f"[DL] TRAIN v{self._train_version} loader_id={id(loader)} ds_id={id(ds)}")
    return loader
```
### DDP Synchronization Flow:
Epoch N End → Callback updates window → All ranks get same window
            ↓
Epoch N+1 Start → PyTorch Lightning detects reload_dataloaders_every_n_epochs=1
                ↓  
Fresh DataLoaders created → All ranks synchronized with NEW objects

### Testing & Validation
✅ Multi-GPU DDP Testing: Verified synchronization across ranks
✅ Object ID Tracking: Confirmed fresh object creation on every reload
✅ Memory Safety: No stale reference detection in extensive testing
✅ Training Stability: Consistent gradients across distributed ranks

### CODE REVIEW

**Comment by mlaurako** [comment]
## OOM Issue:

I lowered `train_window_days` to `12` and decreased the `hidden dim` to `64`. The model was able to run successfully, and from the outputs, I discovered the following:

- The model loads 8 × training_window_days worth of data on Rank 0 for distribution to each GPU.

There are 8 sets of training data for each "version":
```
[DM.set_train_window] -> 2022-10-21 00:00:00 .. 2022-11-02 00:00:00
[DM.set_train_window] v1 -> 2022-10-21 00:00:00 .. 2022-11-02 00:00:00
[DM.set_train_window] -> 2022-05-15 00:00:00 .. 2022-05-27 00:00:00
[DM.set_train_window] v1 -> 2022-05-15 00:00:00 .. 2022-05-27 00:00:00
[DM.set_train_window] -> 2022-07-29 00:00:00 .. 2022-08-10 00:00:00
[DM.set_train_window] v1 -> 2022-07-29 00:00:00 .. 2022-08-10 00:00:00
[DM.set_train_window] -> 2022-05-14 00:00:00 .. 2022-05-26 00:00:00
[DM.set_train_window] v1 -> 2022-05-14 00:00:00 .. 2022-05-26 00:00:00
[DM.set_train_window] -> 2022-11-06 00:00:00 .. 2022-11-18 00:00:00
[DM.set_train_window] v1 -> 2022-11-06 00:00:00 .. 2022-11-18 00:00:00
[DM.set_train_window] -> 2022-11-05 00:00:00 .. 2022-11-17 00:00:00
[DM.set_train_window] v1 -> 2022-11-05 00:00:00 .. 2022-11-17 00:00:00
[DM.set_train_window] -> 2022-04-19 00:00:00 .. 2022-05-01 00:00:00
[DM.set_train_window] v1 -> 2022-04-19 00:00:00 .. 2022-05-01 00:00:00
[DM.set_train_window] -> 2022-01-28 00:00:00 .. 2022-02-09 00:00:00
[DM.set_train_window] v1 -> 2022-01-28 00:00:00 .. 2022-02-09 00:00:00
[DM.set_train_window] -> 2022-04-25 00:00:00 .. 2022-05-07 00:00:00
[DM.set_train_window] v2 -> 2022-04-25 00:00:00 .. 2022-05-07 00:00:00
[DM.set_train_window] -> 2022-05-17 00:00:00 .. 2022-05-29 00:00:00
[DM.set_train_window] v2 -> 2022-05-17 00:00:00 .. 2022-05-29 00:00:00
[DM.set_train_window] -> 2022-09-01 00:00:00 .. 2022-09-13 00:00:00
[DM.set_train_window] v2 -> 2022-09-01 00:00:00 .. 2022-09-13 00:00:00
[DM.set_train_window] -> 2022-05-05 00:00:00 .. 2022-05-17 00:00:00
[DM.set_train_window] v2 -> 2022-05-05 00:00:00 .. 2022-05-17 00:00:00
[DM.set_train_window] -> 2022-10-25 00:00:00 .. 2022-11-06 00:00:00
[DM.set_train_window] v2 -> 2022-10-25 00:00:00 .. 2022-11-06 00:00:00
[DM.set_train_window] -> 2022-05-12 00:00:00 .. 2022-05-24 00:00:00
[DM.set_train_window] v2 -> 2022-05-12 00:00:00 .. 2022-05-24 00:00:00
[DM.set_train_window] -> 2022-10-23 00:00:00 .. 2022-11-04 00:00:00
[DM.set_train_window] v2 -> 2022-10-23 00:00:00 .. 2022-11-04 00:00:00
[DM.set_train_window] -> 2022-10-13 00:00:00 .. 2022-10-25 00:00:00
[DM.set_train_window] v2 -> 2022-10-13 00:00:00 .. 2022-10-25 00:00:00
```

All of these are loaded on Rank 0, causing higher memory requirement:
```
[Rank 0] [Sampler - TRAIN] Next -> 2022-10-21 .. 2022-11-02
[Rank 0] [Sampler - TRAIN] Next -> 2022-05-15 .. 2022-05-27
[Rank 0] [Sampler - TRAIN] Next -> 2022-07-29 .. 2022-08-10
[Rank 0] [Sampler - TRAIN] Next -> 2022-05-14 .. 2022-05-26
[Rank 0] [Sampler - TRAIN] Next -> 2022-11-06 .. 2022-11-18
[Rank 0] [Sampler - TRAIN] Next -> 2022-11-05 .. 2022-11-17
[Rank 0] [Sampler - TRAIN] Next -> 2022-04-19 .. 2022-05-01
[Rank 0] [Sampler - TRAIN] Next -> 2022-01-28 .. 2022-02-09
[Rank 0] [Sampler - TRAIN] Next -> 2022-04-25 .. 2022-05-07
[Rank 0] [Sampler - TRAIN] Next -> 2022-05-17 .. 2022-05-29
[Rank 0] [Sampler - TRAIN] Next -> 2022-09-01 .. 2022-09-13
[Rank 0] [Sampler - TRAIN] Next -> 2022-05-05 .. 2022-05-17
[Rank 0] [Sampler - TRAIN] Next -> 2022-10-25 .. 2022-11-06
[Rank 0] [Sampler - TRAIN] Next -> 2022-05-12 .. 2022-05-24
[Rank 0] [Sampler - TRAIN] Next -> 2022-10-23 .. 2022-11-04
[Rank 0] [Sampler - TRAIN] Next -> 2022-10-13 .. 2022-10-25
```

I've reviewed the code changes, and they appear very similar to my changes. I haven't been able to identify the root cause of this behavior. That said, the randomness isn't necessarily bad to model training. We could also reduce the loaded training window to 2 or 3 days, since each GPU uses approximately 3 pairs of data bins per epoch:

```
gnn_train_3875733.out:[Rank 4] bin4 input_time: 2022-01-02 12:00:00+00:00
**gnn_train_3875733.out:[Rank 0] bin11 input_time: 2022-01-06 00:00:00+00:00**
gnn_train_3875733.out:[Rank 4] bin3 input_time: 2022-01-02 00:00:00+00:00
**gnn_train_3875733.out:[Rank 0] bin6 input_time: 2022-01-03 12:00:00+00:00**
gnn_train_3875733.out:[Rank 4] bin1 input_time: 2022-01-01 00:00:00+00:00
**gnn_train_3875733.out:[Rank 0] bin8 input_time: 2022-01-04 12:00:00+00:00**
gnn_train_3875733.out:[Rank 7] bin10 input_time: 2022-01-05 12:00:00+00:00
gnn_train_3875733.out:[Rank 7] bin18 input_time: 2022-01-09 12:00:00+00:00
gnn_train_3875733.out:[Rank 7] bin9 input_time: 2022-01-05 00:00:00+00:00
gnn_train_3875733.out:[Rank 6] bin10 input_time: 2022-01-05 12:00:00+00:00
gnn_train_3875733.out:[Rank 6] bin2 input_time: 2022-01-01 12:00:00+00:00
gnn_train_3875733.out:[Rank 5] bin9 input_time: 2022-01-05 00:00:00+00:00
gnn_train_3875733.out:[Rank 5] bin22 input_time: 2022-01-11 12:00:00+00:00
gnn_train_3875733.out:[Rank 6] bin4 input_time: 2022-01-02 12:00:00+00:00
gnn_train_3875733.out:[Rank 5] bin8 input_time: 2022-01-04 12:00:00+00:00
gnn_train_3875733.out:[Rank 1] bin23 input_time: 2022-01-12 00:00:00+00:00
gnn_train_3875733.out:[Rank 1] bin3 input_time: 2022-01-02 00:00:00+00:00
gnn_train_3875733.out:[Rank 1] bin2 input_time: 2022-01-01 12:00:00+00:00
gnn_train_3875733.out:[Rank 3] bin2 input_time: 2022-01-01 12:00:00+00:00
gnn_train_3875733.out:[Rank 3] bin16 input_time: 2022-01-08 12:00:00+00:00
gnn_train_3875733.out:[Rank 3] bin21 input_time: 2022-01-11 00:00:00+00:00
gnn_train_3875733.out:[Rank 2] bin9 input_time: 2022-01-05 00:00:00+00:00
gnn_train_3875733.out:[Rank 2] bin21 input_time: 2022-01-11 00:00:00+00:00
gnn_train_3875733.out:[Rank 2] bin6 input_time: 2022-01-03 12:00:00+00:00
```

**Comment by azadeh-gh** [comment]
@mlaurako
### What’s different between our callback versions

- **When the window changes:**
  Mine updates the next window in on_train_epoch_end, so PL picks it up cleanly on the next epoch via reload_dataloaders_every_n_epochs=1.
  Yours (originally) advanced the window at on_train_epoch_start and re-ran setup("fit"). That works but can be heavier/brittle, and—in DDP—needs to run on every rank, not only rank 0, to avoid desync.

- **Validation behavior:**
  Mine keeps val fixed by default (resample is opt-in). That stabilizes early stopping/checkpointing.
  Yours re-sampled val each epoch, which can add variance to val metrics.

- **DDP safety / side effects:**
  Mine doesn’t gate mutations behind is_global_zero, so all ranks stay in sync.
  Yours had rank-0 gating in places, which can lead others to keep the old window.

- **Lightweight data reloads:**
  Mine relies on PL’s native dataloader reload at epoch boundaries instead of calling setup() every epoch.
  
 Functionally our current codepaths are now very similar, but the above are the core intent/behavior differences.
 
###   About the “8 × training_window_days on Rank 0”

Those repeated lines don’t mean rank 0 is loading eight simultaneous windows for everybody:

- In DDP, each rank builds its own DataLoader and reads its own slice; rank 0 doesn’t “fan out” the data to other GPUs.

- The duplicated set_train_window lines in your snippet come from updating multiple internal datasets (e.g., different “versions” v1, v2) plus per-rank logging. The [Sampler - TRAIN] Next -> ... list you pasted looks like successive windows (or multiple datasets) being scheduled—not eight windows resident at once in GPU memory.
- Even if CPU RAM sees extra loader workers/prefetch, that doesn’t explain a CUDA OOM.

### Why the OOM is still happening
This is almost certainly model memory, not the windowing:

- With InteractionNet, mem scales roughly with edges × layers × hidden_dim².
- At 10 layers and hidden_dim=96, we’ve historically not been able to run InteractionNet reliably; >64 has OOM’d for us before. Your observation that dropping to 64 makes it run matches that history.
- The logs showing “each GPU uses ~3 pairs of data bins per epoch” also point away from data volume as the driver of CUDA OOM.
**Recommended path (what tends to work):**
- Keep InteractionNet hidden_dim = 64 and/or reduce num_layers to 6–8.
- Keep TRAIN_WINDOW_DAYS ~ 8–10. Reducing to 2–3 will help CPU I/O but won’t fix GPU OOM if the model is oversized.
